From 975515d9a2ccab6ca78f9d2dd9e916e444f568ae Mon Sep 17 00:00:00 2001
From: Thierry Vignaud <thierry.vignaud@gmail.com>
Date: Mon, 8 Jun 2020 06:50:21 +0200
Subject: [PATCH 216/220] fix build with rpm-4.16

Signed-off-by: Thierry Vignaud <thierry.vignaud@gmail.com>
---
 configure.ac | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/configure.ac b/configure.ac
index eff160b..5d33161 100644
--- a/configure.ac
+++ b/configure.ac
@@ -1887,6 +1887,15 @@ if test x"$rpm_sort_excuse" = x ; then
              [Define to 1 if you have the rpm library.])
 fi
 
+if test x"$LIBRPM" = x ; then
+  # Check for rpm library.
+  AC_CHECK_LIB([rpmio], [rpmvercmp], [],
+               [rpm_sort_excuse="rpmio missing rpmvercmp"])
+   LIBRPM="-lrpmio";
+   AC_DEFINE([HAVE_RPMIO], [1],
+             [Define to 1 if you have the rpm library.])
+fi
+
 AC_SUBST([LIBRPM])
 
 LIBGEOM=
-- 
1.8.3.1

