From 2332d14570ed41a15c36ff264b9790d832106607 Mon Sep 17 00:00:00 2001
From: fengtao <fengtao40@huawei.com>
Date: Tue, 10 Sep 2019 21:43:31 +0800
Subject: [PATCH 220/220] fix grub2-setpassword errors for openEuler

---
 util/grub-set-password.in | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/util/grub-set-password.in b/util/grub-set-password.in
index c0b5ebb..487fbb1 100644
--- a/util/grub-set-password.in
+++ b/util/grub-set-password.in
@@ -1,6 +1,6 @@
 #!/bin/sh -e
 
-EFIDIR=$(grep ^ID= /etc/os-release | sed -e 's/^ID=//' -e 's/rhel/redhat/' -e 's/\"//g')
+EFIDIR=$(grep ^ID= /etc/os-release | sed -e 's/^ID=//' -e 's/rhel/redhat/' -e 's/\"//g' | tr -d '"')
 if [ -d /sys/firmware/efi/efivars/ ]; then
     grubdir=`echo "/@bootdirname@/efi/EFI/${EFIDIR}/" | sed 's,//*,/,g'`
 else
-- 
1.8.3.1

